internal/datastore/crdb/tx.go

```diff
@@ -85,13 +83,26 @@ func execute(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transac
 
 	for i = 0; i < maxRetries; i++ {
 		if err = releasedFn(tx); err != nil {
-			if !retriable(ctx, err) {
-				return err
-			}
-			if _, retryErr := tx.Exec(ctx, "ROLLBACK TO SAVEPOINT cockroach_restart"); retryErr != nil {
-				return fmt.Errorf(errUnableToRetry, err)
+			if retriable(ctx, err) {
+				if _, retryErr := tx.Exec(ctx, "ROLLBACK TO SAVEPOINT cockroach_restart"); retryErr != nil {
+					return fmt.Errorf(errUnableToRetry, err)
+				}
+				continue
+			} else if resetable(ctx, err) {
+				// Close connection and retry with newly acquired connection
+				tx.Rollback(ctx)
+				tx.Conn().Close(ctx)
+				newConn, acqErr := conn.Acquire(ctx)
+				if acqErr != nil {
+					return fmt.Errorf(errUnableToRetry, err)
+				}
+				tx, err = newConn.BeginTx(ctx, txOptions)
```

Comment:
since this is all the same setup as the beginning of the fn, maybe there's a nice way to structure `resets` as a wrapper around `retries`?

Reply:
Hmm yeah, let me try wrapping it and incorporating the above feedback about reset failed retries if possible